src/collection/tx_data/collector.py

```diff
@@ -0,0 +1,282 @@
+"""
+Class for collection of transaction data from Solana chain
+"""
+import asyncio
+import traceback
+from abc import abstractmethod
+from typing import List, Tuple
+import logging
+import os
+import time
+
+from solana.exceptions import SolanaRpcException
+from solana.rpc.async_api import AsyncClient
+from solana.rpc.commitment import Commitment
+from solana.rpc.core import RPCException
+from solders.errors import SerdeJSONError
+from solders.transaction_status import EncodedTransactionWithStatusMeta, UiConfirmedBlock
+from sqlalchemy.exc import IntegrityError
+
+from src.collection.shared.generic_collector import GenericSolanaConnector, SolanaTransaction, log_performance_time
+import db
+
+
+LOG = logging.getLogger(__name__)
+
+
+class TXFromBlockCollector(GenericSolanaConnector):
+    """
+    Class to collect transaction data from Solana blocks.
+    """
+    protocol_public_keys: List[str] | None = None
+
+    def __init__(self):
+        super().__init__()
+        self.async_solana_client = AsyncClient(self.authenticated_rpc_url)
+
+    async def _async_rate_limit_calls(self) -> None:
+        """
+        This method implements a rate-limiting mechanism to control the frequency of API calls asynchronously.
+        It ensures that the number of API calls does not exceed a predefined limit per second.
+        If the limit is reached, it will pause the execution momentarily before allowing further API calls.
+        """
+        # Check if the number of calls made in the last second exceeds the maximum limit.
+        while len(self._call_timestamps) >= self.rate_limit:
+            # Calculate the time passed since the first call in the timestamp list.
+            time_passed = time.time() - self._call_timestamps[0]
+
+            if time_passed > 1:
+                # More than a second has passed since the first call, remove it from the list.
+                self._call_timestamps.pop(0)
+            else:
+                # Wait for the remainder of the second before allowing more calls, asynchronously.
+                await asyncio.sleep(1 - time_passed)
+
+        # Record the timestamp of the current call.
+        self._call_timestamps.append(time.time())
+
+    def run(self):
+        """
+        Main method to process all necessary operations to collect and write data.
+        """
+        raise NotImplementedError("Implement me!")
+
+    async def async_run(self):
```

Comment:
Could we use mixins to add the async versions of the methods?

Reply:
we could but i cant see what the benefits are.